File: exo4.3.py
import os
import sys

# Programme père : crée un tube, crée un fils, transmet le descripteur via argument
# Programme fils (exécuté par execv) : lit le message et l'affiche

# Ceci est le programme principal (père)
def main():
    r, w = os.pipe()
    pid = os.fork()

    if pid == 0:  # Processus fils
        os.close(w)
        # Exécuter le programme fils avec le descripteur de lecture en argument
        # Note : Ce script doit être exécuté avec un programme fils séparé
        # Pour une démonstration autonome, nous allons utiliser une fonction
        os.execl(sys.executable, sys.executable, __file__, "fils", str(r))
    else:  # Processus père
        os.close(r)
        os.write(w, b"Hello exec.\n")
        os.close(w)
        os.wait()

# Programme fils (exécuté quand le script est appelé avec "fils" comme argument)
def fils_programme(fd_lecture):
    buffer = os.read(fd_lecture, 24)
    print(f"Message recu par execv : {buffer.decode()}")
    os.close(fd_lecture)

if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "fils":
        # Exécution en tant que programme fils
        fd_lecture = int(sys.argv[2])
        fils_programme(fd_lecture)
    else:
        main()

File: exo4.3.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>

/*Écrire un programme qui crée un tube, crée un processus ﬁls,
puis, dans le ﬁls, lance par execv un autre programme, appelé programme ﬁls. Le programme
père transmets les descripteurs de tubes au programmes ﬁls en argument, et transmet un
message au ﬁls par le tube. Le programme ﬁls aﬃche le message.*/
int main () {

    int tube[2];

    pipe (tube);

    if (fork () == 0) {
        char arg[16];
        sprintf (arg, "%d", tube[0]);
        char* args[] = {"./fils", arg, NULL};
        close (tube[1]);
        execv (args[0], args);
    } else {
        close (tube[0]);
        write (tube[1], "Hello exec.\n", 11);
        close (tube[1]);
    }
    return 0;
}

//fils.c
int main (int argc, char* argv[]) {
    int fd_lecture = atoi (argv[1]);
    char buffer[24];
    read (fd_lecture, buffer, sizeof (buffer));
    printf ("Message recu par execv : %s\n", buffer);

    return 0;
}